## Import 

In [1]:
import pandas as pd
import scipy.stats as stats
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv('data/kc_house_data.csv')

## Investigate Data

#### Other data to possibly consider: 
* **Census Data of King County** - https://www.census.gov/quickfacts/fact/table/kingcountywashington,WA/PST045219
* **School Quality** - https://educationdata.urban.org/data-explorer/schools/
* **Walk Score** - https://www.walkscore.com/professional/walk-score-apis.php
* **King County GIS Data** - https://gis-kingcounty.opendata.arcgis.com/

In [3]:
data.shape

(21597, 21)

In [4]:
data.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0      5650     1.0        NaN  NONE  ...      7 Average       1180   
1      7242     2.0         NO  NONE  ...      7 Average       2170   
2     10000     1.0         NO  NONE  ...  6 Low Average        770   
3      5000     1.0         NO  NONE  ...      7 Average       1050   
4      8080     1.0         NO  NONE  ...         8 Good       1680   

   sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0            0.0     1955           0.0    98178  47.5112 -122.257   
1          400.0     1951        1991.0    98125  47.7210 -122.319   
2            0.0     1933           NaN    98028  47.7379 -122.233   
3          910.0     1965           0.0    98136  47.5208 -122.393   
4            0.0     1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 21 columns]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

The columns mean: 
* `id` - Unique identifier for a house
* `date` - Date house was sold
* `price` - Sale price (prediction target)
* `bedrooms` - Number of bedrooms
* `bathrooms` - Number of bathrooms
* `sqft_living` - Square footage of living space in the home
* `sqft_lot` - Square footage of the lot
* `floors` - Number of floors (levels) in house
* `waterfront` - Whether the house is on a waterfront
  * Includes Duwamish, Elliott Bay, Puget Sound, Lake Union, Ship Canal, Lake Washington, Lake Sammamish, other lake, and river/slough waterfronts
* `view` - Quality of view from house
  * Includes views of Mt. Rainier, Olympics, Cascades, Territorial, Seattle Skyline, Puget Sound, Lake Washington, Lake Sammamish, small lake / river / creek, and other
* `condition` - How good the overall condition of the house is. Related to maintenance of house.
    * 1 = Poor- Worn out. Repair and overhaul needed on painted surfaces, roofing, plumbing, heating and numerous functional inadequacies. Excessive deferred maintenance and abuse, limited value-in-use, approaching abandonment or major reconstruction 
    * 2 = Fair- Badly worn. Much repair needed. Many items need refinishing or overhauling, deferred maintenance obvious, inadequate building utility and systems all shortening the life expectancy and increasing the effective age.
    * 3 = Average- Some evidence of deferred maintenance and normal obsolescence with age in that a few minor repairs are needed, along with some refinishing. All major components still functional and contributing toward an extended life expectancy. Effective age and utility is standard for like properties of its class and usage.
    * 4 = Good- No obvious maintenance required but neither is everything new. Appearance and utility are above the standard and the overall effective age will be lower than the typical property.
    * 5= Very Good- All items well maintained, many having been overhauled and repaired as they have shown signs of wear, increasing the life expectancy and lowering the effective age with little deterioration or obsolescence evident with a high degree of utility.
* `grade` - Overall grade of the house. Related to the construction and design of the house.
    * Represents the construction quality of improvements. Grades run from grade 1 to 13. Generally defined as:
        * 1-3 Falls short of minimum building standards. Normally cabin or inferior structure.
        * 4 Generally older, low quality construction. Does not meet code.
        * 5 Low construction costs and workmanship. Small, simple design.
        * 6 Lowest grade currently meeting building code. Low quality materials and simple designs.
        * 7 Average grade of construction and design. Commonly seen in plats and older sub-divisions.
        * 8 Just above average in construction and design. Usually better materials in both the exterior and interior finish work.
        * 9 Better architectural design with extra interior and exterior design and quality.
        * 10 Homes of this quality generally have high quality features. Finish work is better and more design quality is seen in the floor plans. Generally have a larger square footage.
        * 11 Custom design and higher quality finish work with added amenities of solid woods, bathroom fixtures and more luxurious options.
        * 12 Custom design and excellent builders. All materials are of the highest quality and all conveniences are present.
        * 13 Generally custom designed and built. Mansion level. Large amount of highest quality cabinet work, wood trim, marble, entry ways etc.
* `sqft_above` - Square footage of house apart from basement
* `sqft_basement` - Square footage of the basement
* `yr_built` - Year when house was built
* `yr_renovated` - Year when house was renovated
* `zipcode` - ZIP Code used by the United States Postal Service
* `lat` - Latitude coordinate
* `long` - Longitude coordinate
* `sqft_living15` - The square footage of interior housing living space for the nearest 15 neighbors
* `sqft_lot15` - The square footage of the land lots of the nearest 15 neighbors

In [6]:
#review how 'view' is coded: 
data['view'].value_counts()

NONE         19422
AVERAGE        957
GOOD           508
FAIR           330
EXCELLENT      317
Name: view, dtype: int64

In [7]:
data.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    sqft_above      yr_built  yr_renovated  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  17755.000000   
mean   1.509941e+04      1.494096   1788.596842   1970.999676     83.636778   
std    4.141264e+04      0.539683    827.759761     29.375234    399.946414   
min    5.200000e+02      1.000000    370.000000   1900.000000      0.000000   
25%    5.040000e+03      1.000000   1190.000000   1951.000000      0.000000   
50%    7.618000e+03      1.500000   1560.000000   1975.000000      0.000000   
75%    1.068500e+04      2.000000   2210.000000   1997.000000      0.000000   
max    1.651359e+06      3.500000   9410.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000  21597.000000   21597.000000   21597.000000  
mean   98077.951845     47.560093   -122.213982    1986.620318   12758.283512  
std       53.513072      0.138552      0.140724     685.230472   27274.441950  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471100   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.231000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [8]:
object_elements = [x for x in data.columns if data[x].dtype=='object']
data[object_elements].describe()

date waterfront   view condition      grade sqft_basement
count       21597      19221  21534     21597      21597         21597
unique        372          2      5         5         11           304
top     6/23/2014         NO   NONE   Average  7 Average           0.0
freq          142      19075  19422     14020       8974         12826

## Investigate null values :
Could nulls have some meaning? For exmaple, are NaN acually 0s?

In [9]:
data['waterfront'].isna().sum()

2376

In [10]:
data.loc[data['waterfront'].isna()].head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0   7129300520  10/13/2014  221900.0         3        1.0         1180   
10  1736800520    4/3/2015  662500.0         3        2.5         3560   
23  8091400200   5/16/2014  252700.0         2        1.5         1070   
40  5547700270   7/15/2014  625000.0         4        2.5         2570   
55  9822700295   5/12/2014  885000.0         4        2.5         2830   

    sqft_lot  floors waterfront  view  ...      grade sqft_above  \
0       5650     1.0        NaN  NONE  ...  7 Average       1180   
10      9796     1.0        NaN  NONE  ...     8 Good       1860   
23      9643     1.0        NaN  NONE  ...  7 Average       1070   
40      5520     2.0        NaN  NONE  ...   9 Better       2570   
55      5000     2.0        NaN  NONE  ...   9 Better       2830   

    sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0             0.0     1955           0.0    98178  47.5112 -122.257   
10         1700.0     1965           0.0    98007  47.6007 -122.145   
23            0.0     1985           NaN    98030  47.3533 -122.166   
40            0.0     2000           NaN    98074  47.6145 -122.027   
55            0.0     1995           0.0    98105  47.6597 -122.290   

    sqft_living15  sqft_lot15  
0            1340        5650  
10           2210        8925  
23           1220        8386  
40           2470        5669  
55           1950        5000  

[5 rows x 21 columns]

In [11]:
data.loc[data['waterfront']=='YES'].head()

id        date      price  bedrooms  bathrooms  sqft_living  \
49    822039084   3/11/2015  1350000.0         3       2.50         2753   
230  8096000060   4/13/2015   655000.0         2       1.75         1450   
246  2025069065   9/29/2014  2400000.0         4       2.50         3650   
264  2123039032  10/27/2014   369900.0         1       0.75          760   
300  3225069065   6/24/2014  3080000.0         4       5.00         4550   

     sqft_lot  floors waterfront       view  ...         grade sqft_above  \
49      65005     1.0        YES    AVERAGE  ...      9 Better       2165   
230     15798     2.0        YES  EXCELLENT  ...     7 Average       1230   
246      8354     1.0        YES  EXCELLENT  ...      9 Better       1830   
264     10079     1.0        YES  EXCELLENT  ...        5 Fair        760   
300     18641     1.0        YES  EXCELLENT  ...  10 Very Good       2600   

     sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
49           588.0     1953           0.0    98070  47.4041 -122.451   
230          220.0     1915        1978.0    98166  47.4497 -122.375   
246         1820.0     2000           0.0    98074  47.6338 -122.072   
264            0.0     1936           0.0    98070  47.4683 -122.438   
300         1950.0     2002           0.0    98074  47.6053 -122.077   

     sqft_living15  sqft_lot15  
49            2680       72513  
230           2030       13193  
246           3120       18841  
264           1230       14267  
300           4550       19508  

[5 rows x 21 columns]

In [12]:
data['view'].value_counts()

NONE         19422
AVERAGE        957
GOOD           508
FAIR           330
EXCELLENT      317
Name: view, dtype: int64

In [13]:
#number nulls yr_renovated column
data['yr_renovated'].isna().sum()

3842

In [14]:
data['yr_renovated'].value_counts()

0.0       17011
2014.0       73
2003.0       31
2013.0       31
2007.0       30
          ...  
1946.0        1
1959.0        1
1971.0        1
1951.0        1
1954.0        1
Name: yr_renovated, Length: 70, dtype: int64

Searched King County Assessory Glossery which did not provide certian of what 0 represents given this data set. In the case of `yr_renovated`, it appears that 0s and NaN can be interpreted as lacking data about a renovation, we will interpret this as "no renovation." 

In [15]:
#change nulls to 0s
data['yr_renovated'].fillna(value=0, inplace=True)

##  Investigate non-numeric data

### sqft_basement
`sqft_basement` is an object type. One would anticipate basement squarefootage to be measured as a number (in square feet). 

In [16]:
data['sqft_basement'].value_counts()

0.0       12826
?           454
600.0       217
500.0       209
700.0       208
          ...  
1525.0        1
946.0         1
792.0         1
276.0         1
768.0         1
Name: sqft_basement, Length: 304, dtype: int64

Homes without a basement have 0. There are 454 '?' Since the question marks do not add informational value to our data set, and  prevent computational work with the data, they will be converted to 0s and the column to a numerical type. 

In [17]:
#replace "?"
data['sqft_basement'].replace(to_replace="?", value=0, inplace=True)

In [18]:
#change type
data['sqft_basement'] = data['sqft_basement'].astype(float)

### Dates
`date`, `yr_built`, and `yr_renovated` are dates within the data set. However they are not indicated as datetime data types. To work with the data appropriately, the columns will be converted to datetime64. 

In [19]:
data['date'] = pd.to_datetime(data['date'], infer_datetime_format=True)
data['date'] = data['date'].dt.date

In [20]:
data['date'][1]

datetime.date(2014, 12, 9)

What are the sales dates of the dataset? 

In [21]:
min_saledate = data['date'].min()
max_saledate = data['date'].max()

print(f' This data set contains homes sold from {min_saledate} to {max_saledate}.')

 This data set contains homes sold from 2014-05-02 to 2015-05-27.


### Locations

`zipcode`, `lat`, and `long` are location-based data. Using this information we may be able to gleam other insights about the sales prices of houses using a home's location relative to other places. Some questions to investigate: 
* Can zipcodes help predict prices? 
* Does the location of a home to the city-center impact home price? 
* Does the location of a home to city-services impact sales price (independent of location to city-ceter)?
    * Services to consider: shcools, parks, transportation, grocery stores
       * School Quality - https://educationdata.urban.org/data-explorer/schools/
       * Walk Score - https://www.walkscore.com/professional/walk-score-apis.php

In [22]:
#Number of zipcodes represented in data set
len(data['zipcode'].value_counts())

70

Zip Code List - https://www.ciclt.net/sn/clt/capitolimpact/gw_ziplist.aspx?FIPS=53033

In [23]:
#create "coordinates" column for mapping
data['coordinates'] = data[['lat', 'long']].values.tolist()

In [24]:
# If we decide to do mapping 

#import numpy as np
#import folium 

# houses_map = folium.Map(location=[47.6039457,-122.3298889],
#                         tiles='OpenStreetMap', zoom_start=10)

# houses = np.random.choice(data['coordinates'], size=10)
# houses_loc = list(houses)

# for house in range(0, len(houses_loc)): 
#       folium.Marker(house_loc[house]).add_to(houses_map)

In [25]:
#houses_map

# Feature Engineering
## Ratios
There are several columns that may be benificial to combine and review as ratios. 

* **Bedrooms : Bathrooms** - values close to one show a balance of bedrooms to bathrooms. Values less than one show more bathrooms than bedrooms (strange house). 

* **Lot square footage: Home square footage** - help determine yard size 

* **Home square footage : Neighbor sqarefootage**  - compare house size to the neighbors' houses

* **Squarefootage in terms of number of bedrooms** - Is there a blance of bedrooms and size of the house?

In [26]:
def ratios(data, num, denom):
    column = num + "_" + denom
    data[column] = data[num]/data[denom]

ratios(data, 'bedrooms', 'bathrooms')
ratios(data, 'sqft_lot', 'sqft_living')
ratios(data, 'sqft_living', 'sqft_living15')
ratios(data, 'sqft_living', 'bedrooms')

## Categorical Data

* Grade (using the language from the documentation, we can categorize these further into high-medium-low sub categories. Houses <=6 all usesd terms such as "low", houses with grades between 7 and 10 all used terms such as "average", and the >10 grades were all "high-quality."

In [27]:
def num_grade(row): 
    return row['grade'].split()[0]

In [28]:
data['grade'] = data.apply(lambda row: num_grade(row), axis=1).astype(int)

In [29]:
def grade_category(row):
    if row['grade'] <=6 :
        return 'low'
    elif row['grade']>6 and row['grade']<10:
        return 'medium'
    else:
        return 'high'

In [30]:
data['grade_category'] = data.apply(lambda row: grade_category(row), axis=1).astype("category")

* Condition is another categorical variable. 

In [31]:
data['condition'] = data['condition'].astype('category')

# Current State of the data:

In [32]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 27 columns):
 #   Column                     Non-Null Count  Dtype   
---  ------                     --------------  -----   
 0   id                         21597 non-null  int64   
 1   date                       21597 non-null  object  
 2   price                      21597 non-null  float64 
 3   bedrooms                   21597 non-null  int64   
 4   bathrooms                  21597 non-null  float64 
 5   sqft_living                21597 non-null  int64   
 6   sqft_lot                   21597 non-null  int64   
 7   floors                     21597 non-null  float64 
 8   waterfront                 19221 non-null  object  
 9   view                       21534 non-null  object  
 10  condition                  21597 non-null  category
 11  grade                      21597 non-null  int32   
 12  sqft_above                 21597 non-null  int64   
 13  sqft_basement              2159

## The aim is to create a model for home *sellers*

This model will aim to identify what features of a home that will incrase the resell value, and by what amount. 

Therefore, we will keep variables that are associated with selling a home, and drop variables that homeowners cannot change (for example, location based data such as waterfront property).

In [33]:
columns = ['id', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated','sqft_living15', 'sqft_lot15', 'coordinates',
       'bedrooms_bathrooms', 'sqft_lot_sqft_living',
       'sqft_living_sqft_living15', 'sqft_living_bedrooms', 'grade_category']

data = data[columns]